lib/iris/aux_factory.py

```diff
@@ -784,20 +758,51 @@ def dependencies(self):
                     depth_c=self.depth_c, nsigma=self.nsigma, zlev=self.zlev)
 
     def _derive(self, sigma, eta, depth, depth_c,
-                zlev, shape, nsigma_slice):
+                zlev, nsigma, coord_dims_func):
+        # Calculate the index of the 'z' dimension in the inputs.
+        # Get the cube dimension...
+        i_levels_cubedim, = coord_dims_func(self.dependencies['zlev'])
+        # ...subtract any previous dimensions not used in the calculation.
+        derived_cubedims = self.derived_dims(coord_dims_func)
+        i_levels_dim = i_levels_cubedim - sum(
+            i_dim not in derived_cubedims
+            for i_dim in range(i_levels_cubedim))
+
+        # Calculate the result shape as a combination of all the inputs.
+        allshapes = np.array(
+            [el.shape
+             for el in (sigma, eta, depth, depth_c, zlev)
+             if el.ndim])
+        result_shape = list(np.max(allshapes, axis=0))
+        ndims = len(result_shape)
+        # Make a slice tuple to index the first nsigma z-levels.
+        z_slices_nsigma = [slice(None)] * ndims
+        z_slices_nsigma[i_levels_dim] = slice(0, int(nsigma))
+        z_slices_nsigma = tuple(z_slices_nsigma)
+        # Make a slice tuple to index the remaining z-levels.
+        z_slices_rest = [slice(None)] * ndims
+        z_slices_rest[i_levels_dim] = slice(int(nsigma), None)
+        z_slices_rest = tuple(z_slices_rest)
         # Perform the ocean sigma over z coordinate nsigma slice.
         if eta.ndim:
-            eta = eta[nsigma_slice]
+            eta = eta[z_slices_nsigma]
         if sigma.ndim:
-            sigma = sigma[nsigma_slice]
+            sigma = sigma[z_slices_nsigma]
         if depth.ndim:
-            depth = depth[nsigma_slice]
+            depth = depth[z_slices_nsigma]
         # Note that, this performs a point-wise minimum.
-        temp = eta + sigma * (np.minimum(depth_c, depth) + eta)
-        # Calculate the final derived result.
-        result = np.ones(shape, dtype=temp.dtype) * zlev
-        result[nsigma_slice] = temp
-
+        nsigma_levs = eta + sigma * (da.minimum(depth_c, depth) + eta)
+        # Expand to full shape, as it may sometimes have lower dimensionality.
+        ones_full_result = np.ones(result_shape, dtype=np.int16)
```

Comment:
> shouldn't this be da.ones(...) ? To keep all things lazy

I thought I'd only use dask where needed, and I thought it wasn't here because "nsigma_levs" is always lazy,  so result is "dask * numpy".  However, you just reminded me that this could be creating a large real array, so I'll change !!